src/Controller/Security/ResetPasswordController.php

```diff
@@ -0,0 +1,170 @@
+<?php
+
+namespace App\Controller\Security;
+
+use App\Entity\User;
+use App\Form\ChangePasswordForm;
+use App\Form\ForgotPasswordForm;
+use App\Repository\UserRepository;
+use Doctrine\ORM\EntityManagerInterface;
+use Symfony\Bridge\Twig\Mime\TemplatedEmail;
+use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
+use Symfony\Bundle\SecurityBundle\Security;
+use Symfony\Component\DependencyInjection\Attribute\Target;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpFoundation\UriSigner;
+use Symfony\Component\Mailer\Header\MetadataHeader;
+use Symfony\Component\Mailer\Header\TagHeader;
+use Symfony\Component\Mailer\MailerInterface;
+use Symfony\Component\PasswordHasher\Hasher\UserPasswordHasherInterface;
+use Symfony\Component\RateLimiter\RateLimiterFactory;
+use Symfony\Component\Routing\Attribute\Route;
+use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
+
+use function Symfony\Component\Clock\now;
+
+final class ResetPasswordController extends AbstractController
+{
+    private const EXPIRES_AFTER = 3600; // 1 hour
+    private const SESSION_ID = 'password_reset_id';
+    private const ALREADY_USED_HASH_PARAM = 'valid';
+
+    public function __construct(
+        private UriSigner $uriSigner,
+        private MailerInterface $mailer,
+        private UserRepository $users,
+        private UserPasswordHasherInterface $passwordHasher,
+        private EntityManagerInterface $em,
+        private Security $security,
+    ) {
+    }
+
+    #[Route('/forgot-password', name: 'password_reset_request')]
+    public function forgot(
+        Request $request,
+
+        #[Target('forgotPasswordEmailLimiter')]
+        RateLimiterFactory $rateLimiter, // todo switch to RateLimiterInterface in 7.3
+    ): Response {
+        $form = $this->createForm(ForgotPasswordForm::class);
+        $form->handleRequest($request);
+
+        if ($form->isSubmitted() && $form->isValid()) {
+            /** @var string $email */
+            $email = $form->get('email')->getData();
+
+            if (!$rateLimiter->create($email)->consume()->isAccepted()) {
+                $this->addFlash('warning', 'You recently requested a password reset. Please check your email for the link to reset your password.');
+
+                return $this->redirectToRoute('password_reset_request');
+            }
+
+            $this->sendResetLink($email);
+
+            $this->addFlash('note', \sprintf('A password reset link has been sent to %s.', $email));
```

Comment:
Do we want to get opinionated to the `note` vs `success` keys? Or just have `success` and `error`?